pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -414,18 +414,28 @@ private void updateBundleData() {
 
             // Remove all loaded bundles from the preallocated maps.
             final Map<String, BundleData> preallocatedBundleData = brokerData.getPreallocatedBundleData();
-            // Should not iterate with more than one thread at a time.
             synchronized (preallocatedBundleData) {
-                final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
-                        .iterator();
-                while (preallocatedIterator.hasNext()) {
-                    final String bundle = preallocatedIterator.next().getKey();
-                    if (bundleData.containsKey(bundle)) {
-                        preallocatedIterator.remove();
-                        preallocatedBundleToBroker.remove(bundle);
+                for (String bundleName : brokerData.getLocalData().getBundles()) {
+                    if (preallocatedBundleData.containsKey(bundleName)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Preallocated bundle {[]} is now loaded on broker [{}]", bundleName, broker);
+                        }
+                        final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
+                                .iterator();
+                        while (preallocatedIterator.hasNext()) {
```

Comment:
> Good point, I think that clearing the map based on ZK watches would be the best option.

Yes, on broker data-change or broker-node removal, zk-watch listener [updates broker-data](https://github.com/yahoo/pulsar/blob/master/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java#L381) where we can cleanup the bundles if broker has been removed.